### PR TITLE
Added support for Qute `@TemplateEnum` annotation

### LINKED ISSUE #485: Support for `@TemplateData`
To main work is to generate on JDT LS side the proper value resolver information like with [@TemplateExtension support](https://github.com/redhat-developer/quarkus-ls/blob/master/qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateExtensionSupport.java) by scanning

* `@TemplateData` annotation ->  https://quarkus.io/guides/qute-reference#template_data
* `@TemplateEnum` annotation -> https://quarkus.io/guides/qute-reference#convenient-annotation-for-enums

to collect the methods/fields/enum and generate value resolver according the https://quarkus.io/guides/qute-reference#template_data

After that, completion, validation, hover, definition, etc will work out of the box in the Qute template.

### PR DESCRIPTION
**Pull request by AlexXuChen**
Added support for Qute `@TemplateEnum` annotation
Peek 2022-05-12 10-55

Fixes #485 

Signed-off-by: Alexander Chen <alchen@redhat.com>

### CODE REVIEW

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+ * @see https://quarkus.io/guides/qute-reference#convenient-annotation-for-enums
+ *
+ */
+public class TemplateEnumAnnotationSupport extends AbstractAnnotationTypeReferenceDataModelProvider {
+
+	private static final Logger LOGGER = Logger.getLogger(TemplateExtensionSupport.class.getName());
+
+	private static final String[] ANNOTATION_NAMES = {
+			TEMPLATE_ENUM_ANNOTATION
+
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java:56]: remove empty line

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	@Override
+	protected String[] getAnnotationNames() {
+		return ANNOTATION_NAMES;
+	}
+
+	@Override
+	protected void processAnnotation(IJavaElement javaElement, IAnnotation annotation, String annotationName,
+			SearchContext context, IProgressMonitor monitor) throws JavaModelException {
+		if (!(javaElement instanceof IAnnotatable)) {
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java:67]: Remplace like TemplateDataAnnotationSupport:

```java
if (javaElement.getElementType() != IJavaElement.TYPE) {
	return;
}
IType type = (IType) javaElement;
IAnnotation templateEnum = AnnotationUtils.getAnnotation((IAnnotatable) javaElement, TEMPLATE_ENUM_ANNOTATION);
if (templateEnum == null) {
	return;
}
```

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	}
+
+	private static void collectResolversForTemplateEnum(IType type, IAnnotation templateExtension,
+			List<ValueResolverInfo> resolvers, IProgressMonitor monitor) {
+		try {
+			ITypeResolver typeResolver = QuteSupportForTemplate.createTypeResolver(type);
+			if (type.isEnum()) {
+				IMethod[] methods = type.getMethods();
+				for (IMethod method : methods) {
+					IAnnotation methodTemplateExtension = AnnotationUtils.getAnnotation(method,
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java:88]: The enum field cannot be annotated with @TemplateEnum, please remove this code.

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+		if (javaElement.getElementType() != IJavaElement.TYPE) {
+			return;
+		}
+		IType type = (IType) javaElement;
+		IAnnotation templateEnum = AnnotationUtils.getAnnotation((IAnnotatable) javaElement, TEMPLATE_ENUM_ANNOTATION);
+		if (templateEnum == null) {
+			return;
+		}
+
+		if (javaElement instanceof IType) {
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java:75]: Why you did that? At this Stéphanie you have the proper IType

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+	}
+
+	@Override
+	protected void processAnnotation(IJavaElement javaElement, IAnnotation annotation, String annotationName,
+			SearchContext context, IProgressMonitor monitor) throws JavaModelException {
+		if (javaElement.getElementType() != IJavaElement.TYPE) {
+			return;
+		}
+		IType type = (IType) javaElement;
+		IAnnotation templateEnum = AnnotationUtils.getAnnotation((IAnnotatable) javaElement, TEMPLATE_ENUM_ANNOTATION);
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java:70]: Check here if type.isEnum

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+		if (templateEnum == null) {
+			return;
+		}
+
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java:74]: Check here if the type is znnotated with @TemplateData in this case ignore by return the value resolver generator

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+	}
+
+	private static void collectResolversForTemplateEnum(IType type, IAnnotation templateExtension,
+			List<ValueResolverInfo> resolvers, IProgressMonitor monitor) {
+		try {
+			ITypeResolver typeResolver = QuteSupportForTemplate.createTypeResolver(type);
+			if (type.isEnum()) {
+				IField[] fields = type.getFields();
+				for (IField field : fields) {
+					IAnnotation fieldFromTemplateEnum = AnnotationUtils.getAnnotation(field, TEMPLATE_ENUM_ANNOTATION);
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java:89]: Is field can be annotated with@TemplateEnum ?

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+		ITypeResolver typeResolver = QuteSupportForTemplate.createTypeResolver(field);
+		collectResolversForTemplateEnum(field, templateExtension, resolvers, typeResolver);
+	}
+
+	private static void collectResolversForTemplateEnum(IField field, IAnnotation templateExtension,
+			List<ValueResolverInfo> resolvers, ITypeResolver typeResolver) {
+		ValueResolverInfo resolver = new ValueResolverInfo();
+		resolver.setSourceType(field.getDeclaringType().getFullyQualifiedName());
+		resolver.setSignature(typeResolver.resolveFieldSignature(field));
+		resolver.setNamespace(field.getDeclaringType().getFullyQualifiedName().replace('.', '_'));
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java:111]: According the qute référence guide the namespace is just the simple  class name.

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java (hunk 7 above)
Comment (AlexXuChen) [reply qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java:89]: As per the reference guide example: 
```java
public enum Status {
    ON,
    OFF
}
```
What are `ON` and `OFF` considered?

Comment (angelozerr) [reply qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java:89]: ON and OFF are fields. I think you cannot write:

@TemplateEnum
ON

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java (hunk 6 above)
Comment (angelozerr) [reply qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java:74]: Remove the next code of this method.

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+		}
+
+		if (javaElement instanceof IType) {
+			IType type = (IType) javaElement;
+			collectResolversForTemplateEnum(type, templateEnum, context.getDataModelProject().getValueResolvers(),
+					monitor);
+		}
+	}
+
+	private static void collectResolversForTemplateEnum(IType type, IAnnotation templateExtension,
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java:82]: Remove the unused parameter `IAnnotation templateExtension`

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+			List<ValueResolverInfo> resolvers, IProgressMonitor monitor) {
+		try {
+			ITypeResolver typeResolver = QuteSupportForTemplate.createTypeResolver(type);
+			if (type.isEnum()) {
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java:86]: Remove this test (it is done before)

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+							fieldFromTemplateEnum != null ? fieldFromTemplateEnum : templateExtension, resolvers,
+							typeResolver);
+				}
+			}
+		} catch (JavaModelException e) {
+			LOGGER.log(Level.SEVERE, "Error while getting methods of '" + type.getElementName() + "'.", e);
+		}
+	}
+
+	public static void collectResolversForTemplateEnum(IField field, IAnnotation templateExtension,
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java:100]: remove this method. the ITypeResolver must be created once time for the type.

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+			List<ValueResolverInfo> resolvers, IProgressMonitor monitor) {
+		ITypeResolver typeResolver = QuteSupportForTemplate.createTypeResolver(field);
+		collectResolversForTemplateEnum(field, templateExtension, resolvers, typeResolver);
+	}
+
+	private static void collectResolversForTemplateEnum(IField field, IAnnotation templateExtension,
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java:106]: Remove the unused  `IAnnotation templateExtension` parameter.

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java (hunk 7 above)
Comment (angelozerr) [reply qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateEnumAnnotationSupport.java:89]: In other words replace with this code:

```
IField[] fields = type.getFields();
for (IField field : fields) {
    collectResolversForTemplateEnum(field, resolvers, typeResolver) {
}
```

**Comment by angelozerr** [comment]
Your PR looks good, it misses just the tests :)

**Comment by angelozerr** [comment]
Good job @AlexXuChen !

**Pull request merged**